File: src/csharp/Intel/Iced/Intel/FastStringOutput.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if FAST_FMT
using System;
using System.Runtime.CompilerServices;

namespace Iced.Intel {
	/// <summary>
	/// <see cref="FastFormatter"/> output
	/// </summary>
	public sealed class FastStringOutput {
		char[] buffer;
		int bufferLen;

		/// <summary>
		/// Gets the current length
		/// </summary>
		public int Length => bufferLen;

		/// <summary>
		/// Constructor
		/// </summary>
		public FastStringOutput() => buffer = new char[64];

		/// <summary>
		/// Constructor
		/// </summary>
		/// <param name="capacity">Initial capacity</param>
		public FastStringOutput(int capacity) => buffer = new char[capacity];

		/// <summary>
		/// Append a char
		/// </summary>
		/// <param name="c">Character to append</param>
		public void Append(char c) {
			var buffer = this.buffer;
			var bufferLen = this.bufferLen;
			if ((uint)bufferLen >= (uint)buffer.Length) {
				Resize(1);
				buffer = this.buffer;
			}
			buffer[bufferLen] = c;
			this.bufferLen = bufferLen + 1;
		}

		/// <summary>
		/// Append a string
		/// </summary>
		/// <param name="value">String to append</param>
		public void Append(string? value) {
			if (value is not null)
				AppendNotNull(value);
		}

		internal void AppendNotNull(string value) {
			var buffer = this.buffer;
			var bufferLen = this.bufferLen;
			if ((uint)bufferLen + value.Length > (uint)buffer.Length) {
				Resize(value.Length);
				buffer = this.buffer;
			}
			for (int i = 0; i < value.Length; i++) {
				buffer[bufferLen] = value[i];
				bufferLen++;
			}
			this.bufferLen = bufferLen;
		}

		[MethodImpl(MethodImplOptions.NoInlining)]
		void Resize(int extraCount) {
			int capacity = buffer.Length;
			int newCount = checked(Math.Max(capacity * 2, capacity + extraCount));
			Array.Resize(ref buffer, newCount);
		}

#if HAS_SPAN
		/// <summary>
		/// Returns the current string as a span. The return value is valid until this instance gets mutated.
		/// </summary>
		/// <returns></returns>
		public ReadOnlySpan<char> AsSpan() =>
			new ReadOnlySpan<char>(buffer, 0, bufferLen);
#endif

		/// <summary>
		/// Copies all data to <paramref name="array"/>
		/// </summary>
		/// <param name="array">Destination array</param>
		/// <param name="arrayIndex">Destination array index</param>
		public void CopyTo(char[] array, int arrayIndex) =>
			Array.Copy(buffer, 0, array, arrayIndex, bufferLen);

		/// <summary>
		/// Resets the buffer to an empty string
		/// </summary>
		public void Clear() => bufferLen = 0;

		/// <summary>
		/// Gets the current string
		/// </summary>
		/// <returns></returns>
		public override string ToString() => new string(buffer, 0, bufferLen);
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/fmt/fast/FastStringOutput.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.fmt.fast;

/**
 * {@link FastFormatter} output
 */
public final class FastStringOutput {
	private char[] buffer;
	private int bufferLen;

	/**
	 * Gets the current length
	 */
	public int size() {
		return bufferLen;
	}

	/**
	 * Constructor
	 */
	public FastStringOutput() {
		buffer = new char[64];
	}

	/**
	 * Constructor
	 *
	 * @param capacity Initial capacity
	 */
	public FastStringOutput(int capacity) {
		buffer = new char[capacity];
	}

	/**
	 * Append a char
	 *
	 * @param c Character to append
	 */
	public void append(char c) {
		char[] buffer = this.buffer;
		int bufferLen = this.bufferLen;
		if (Integer.compareUnsigned(bufferLen, buffer.length) >= 0) {
			resize(1);
			buffer = this.buffer;
		}
		buffer[bufferLen] = c;
		this.bufferLen = bufferLen + 1;
	}

	/**
	 * Append a string
	 *
	 * @param value String to append (may be {@code null})
	 */
	public void append(String value) {
		if (value != null) {
			char[] buffer = this.buffer;
			int bufferLen = this.bufferLen;
			if (Integer.compareUnsigned(bufferLen + value.length(), buffer.length) > 0) {
				resize(value.length());
				buffer = this.buffer;
			}
			for (int i = 0; i < value.length(); i++) {
				buffer[bufferLen] = value.charAt(i);
				bufferLen++;
			}
			this.bufferLen = bufferLen;
		}
	}

	private void resize(int extraCount) {
		int capacity = buffer.length;
		long required = (long)capacity + (long)extraCount;
		long newCount = Math.min(Math.max((long)capacity << 1, required), 0x7FFF_FFFF);
		if (newCount < required)
			throw new OutOfMemoryError();
		char[] newArray = new char[(int)newCount];
		System.arraycopy(buffer, 0, newArray, 0, bufferLen);
		this.buffer = newArray;
	}

	/**
	 * Resets the buffer to an empty string
	 */
	public void clear() {
		bufferLen = 0;
	}

	/**
	 * Gets the current string
	 */
	@Override
	public String toString() {
		return new String(buffer, 0, bufferLen);
	}
}
